Lasso model

jzmujdzin opened the pull request.

jzmujdzin committed:
add lasso notebook

jzmujdzin committed:
add garbage collector for lasso notebook

jzmujdzin committed:
add lasso model

kuba1302 commented on ticket_upgrade_prediction/__init__.py:
```diff
@@ -1,2 +1,2 @@
-from data_prep import Pipeline
-from evaluator import Evaluator
```
Why is it commented out?

jzmujdzin replied:
those imports cause problems when importing packages on my side
`Traceback (most recent call last):
  File "E:/wec_2022/ticket_upgrade_prediction/models/lasso/lasso.py", line 9, in <module>
    from ticket_upgrade_prediction.models.base import BaseModel
  File "E:\wec_2022\ticket_upgrade_prediction\__init__.py", line 1, in <module>
    from pipeline import Pipeline
ModuleNotFoundError: No module named 'pipeline'`

kuba1302 replied:
the easiest solve to it, is to use setup.py file from my net PR, and in main directory, use: 
`pip install -e .`

kuba1302 replied:
I forgot to mention it, sorry

kuba1302 commented on ticket_upgrade_prediction/models/lasso/lasso.py:
```diff
@@ -0,0 +1,79 @@
+    def predict_proba(self, X) -> np.ndarray:
+        return self.model.predict_proba(X)
+
+    @staticmethod
+    def fit_model(
+        dataset: pd.DataFrame,
+        class_weight_balance: str,
+        verbose: int,
+        max_iter: int,
+        target: str,
+    ) -> LogisticRegression:
+        model = LogisticRegression(
+            penalty="l1",
+            solver="saga",
+            class_weight=class_weight_balance,
+            verbose=verbose,
+            max_iter=max_iter,
+        )
+        dataset.dropna(inplace=True)
+        X_train, _, y_train, _ = train_test_split(
+            dataset.drop([target], axis=1),
+            dataset[[target]],
+            test_size=0.2,
+            random_state=42,
+        )
+        del dataset
+        gc.collect()
+        model.fit(X_train, y_train)
+        return model
```
Wouldn't it be better to make it normal method, and do somoething like: 
`self.model = model`?

kuba1302 replied:
If it's static here, i think base class would need to have `abstractstaticmethod` decorator

jzmujdzin replied:
used self.model = None and then applying changes to self.model inside functions

kuba1302 commented on ticket_upgrade_prediction/models/lasso/lasso.py:
```diff
@@ -0,0 +1,79 @@
+        return model
+
+    @staticmethod
+    def get_fitted_model(
```
Same, here, we could have property, `self.model_info`

jzmujdzin replied:
changed it to overwrite self.model

kuba1302 commented on ticket_upgrade_prediction/models/lasso/lasso.py:
```diff
@@ -0,0 +1,79 @@
+            )
+            if from_mlflow
+            else self.fit_model(
+                dataset=kwargs.pop("dataset", Pipeline().df),
+                class_weight_balance=kwargs.pop(
+                    "class_weight_balance", "balanced"
+                ),
+                verbose=kwargs.pop("verbose", 2),
+                max_iter=kwargs.pop("max_iter", 333),
+                target=kwargs.pop("target", "UPGRADED_FLAG"),
+            )
```
Why do we use `dict.pop(<key>, <base_value>)` instead of `dict.get(<key>, <base_value>`?

kuba1302 replied:
Save above :arrow_double_up:

jzmujdzin replied:
sample code when using get:
```
def inside_function(arg1 = 'default_arg_1', arg2 = 'default_arg_2'):
    return arg1, arg2
    
def outer_func(**kwargs):
    print(inside_function(kwargs.get('argument1'), kwargs.get('argument2')))
    
outer_func(argument1='test')
('test', None)
```
dict.pop(key, default_val) tries to return the value from key, and if there's no key it return the default value. this way i ommit passing None argument inside the function

kuba1302 commented on ticket_upgrade_prediction/models/lasso/lasso.py:
```diff
@@ -0,0 +1,79 @@
+from ticket_upgrade_prediction.pipeline import Pipeline
+
+
+class LassoModel(BaseModel):
```
Maybe we should add `save_model()` method to our interface?

kuba1302 replied:
So that we ensure, that every model has save option

jzmujdzin replied:
i.e. save it to a pickle ?

kuba1302 replied:
pickle and another one for mlflow?

kuba1302 commented on ticket_upgrade_prediction/models/lasso/lasso.py:
```diff
@@ -0,0 +1,79 @@
+
+
+class LassoModel(BaseModel):
+    def __init__(self, from_mlflow: bool = False, **kwargs):
```
Maybe lets iniciate it with `self.model = None`, and then depending on situation, use `.get_fitted_model(run_id=<run_id>)` /`.get_fitted_model(model_version=<version>, model_name=<model_name>),`  if we want to download model from mlflow, and `fit_model(self, **kwargs)` if we want to fit model.

In such a way, we can pass diffrent sets of arguments in each method, and it becomes much more readable, then using one `kwargs` dictionary from `__init__`  to supply arguments for two separate methods.

kuba1302 replied:
That of course involves changes in BaseModel interface

jzmujdzin replied:
not sure if i got it right, please check my last commit

kuba1302 replied:
I showed what i mean here: 
https://github.com/kuba1302/ticket-upgrade-prediction/pull/11

kuba1302 commented on ticket_upgrade_prediction/models/lasso/lasso.py:
```diff
@@ -0,0 +1,79 @@
+            test_size=0.2,
+            random_state=42,
+        )
+        del dataset
+        gc.collect()
```
Why do we need this? I dont really know what `gc` library does, and i'm curious about it. Doesn't python clean memory by itself after the function call is finished? Or am i wrong?

jzmujdzin replied:
if you load the dataset in i.e. main, there are two duplicate versions of it if im not wrong. garbage collector deletes the dataset from memory, clearing some RAM space
in case of large dataset (i.e. wec 2k22), i needed to clear memory in some cases (especially jupyter lab) in order to properly use python

kuba1302 commented on ticket_upgrade_prediction/models/lasso/lasso.py:
```diff
@@ -0,0 +1,79 @@
+        self.model = (
+            self.get_fitted_model(
+                tracking_uri=kwargs.pop(
+                    "tracking_uri", "http://localhost:5000"
```
Easiest way to pass tracking uri, is to use `MLFLOW_TRACKING_URI` env variable. 
I forgot to create example env, i will do it later

jzmujdzin replied:
added

jzmujdzin committed:
add mlflow_traing_uri to example env

jzmujdzin committed:
change LassoModel initialization

now it initializes with self.model = None
and then operates on this self.model, changing its attributes
based on whether we want to download the model from mlflow

kuba1302 committed:
change lasso model class

jzmujdzin committed:
Merge pull request #11 from kuba1302/lasso_model-init-fix

Change lasso model class

jzmujdzin committed:
add model saving to .pkl and mlflow

jzmujdzin committed:
black formatting

jzmujdzin committed:
black formatting

jzmujdzin committed:
isort formatting
